### DISCUSSION TITLE
Dynamic DICOM Endpoints.

### DISCUSSION THREAD

**Discussion by dipak101505** [KathiraveluLab/Diomede#4] [General]
Hi @pradeeban !
I'd like to discuss the project's problem statement to ensure we're on the same page.

My understanding:

The project aims to route DICOM files from the source to dynamic endpoints.

Implementation Proposal:

I propose building a gateway and dicom acceptor that can:
-Gateway accepts dicom from source. It has a list of all the endpoints stored in a database.
-The gateway then dynamically routes the DICOM to any one of the acceptor
-Dicomacceptor uploads the dicom to the cloud storage.
-Each dicom acceptor is configured with PACS endpoint and sends the cloud storage object to PACS.

Is my understanding of the project goals correct? Does my proposed approach seem feasible? If so, may I proceed with working on it?

I've also attached a diagram for your reference, providing a high-level overview of the proposed solution.
![dynamic](https://github.com/KathiraveluLab/Diomede/assets/26256895/eee24d08-7ea7-4958-b87e-a5574cf786da)

**Comment by pradeeban** [comment]
@dipak101505, these additional DICOM receiver components will add more latency and queuing delays. I was thinking more along the lines of just having the 3 targets and sources (either the scanners or a PACS).

If it is a PACS, providing the ability to change the destination.

While changing the destination dynamically in pre-existing PACS will be tough, our own DICOM senders can be configured to do that. This was my thought process.

**Reply by dipak101505** [reply]
Just to be on the same page, are you proposing to build a dicom sender application that receives dicom from scanners and routes it to dynamic endpoints?

**Reply by pradeeban** [reply]
There are several already existing DICOM sender applications such as https://github.com/Emory-HITI/Niffler

So, no need to reinvent the wheel. The proposal is the dynamic nature of destination. DICOM expects the destination to be static. So an A/B proxy style change of destination is not feasible, just by following the protocol.

What we need is an extension to existing DICOM receivers (such as Niffler, Orthanc, etc). Perhaps a minimal proxy that intercepts and routes the DICOM images to the actual DICOM endpoint? 

I would imagine a simplistic approach. You are free to have your own design, of course. There are always several ways to achieve the dynamic destination feature, within the bounds of the DICOM protocol.

**Comment by Nitesh639** [comment]
Hi @pradeeban,
I'm Nitesh Kumar Kushwaha, a GSoC contributor in 2022 at Emory BMI under your mentorship. I'm particularly drawn to the "Dynamic DICOM Endpoints" project for Google Summer of Code.

I'm eager to re-engage with this community. Proficient in Python and Java, I bring experience in handling DICOM images, a skill vital for this project and others such as 'Creating shareable "albums" from locally stored DICOM images' and 'DICOM Image Retrieval and Processing in Matlab'.

Hence, I'm keen to contribute to these projects and collaborate with fellow community members.
